### PR TITLE
Rename the plugin oneapi->sycl

### PR DESCRIPTION
**Pull request by razdoburdin**
List of the provided changes:
- The plugin and all naming inside it change "`oneapi`->`sycl`".
- Added new namespace `sycl `with nested namespaces copying the structure of main namespaces. Functions are renamed: "`xgboost::<name_space>::<FunctionNameOneAPI>` -> `xgboost::sycl::<name_space>::<FunctionName>`.
- Previously, plugin addressed the calls to CPU  implementation if device is set to `sycl:cpu`. This PR change this behavior. Now for `device` is `sycl:cpu` SYCL implementation is launched on available `sycl:cpu` device.
- Warnings not related to the plugin are suspended.
- Tests for SYCL implementation of multi:softmax and multi:softprob were added.

### CODE REVIEW

File: tests/cpp/objective/test_multiclass_obj.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+void TestSoftmaxMultiClassObjGPair(const Context* ctx);
+
+void TestSoftmaxMultiClassBasic(const Context* ctx);
+
+void TestSoftprobMultiClassBasic(const Context* ctx);
+
+}  // namespace xgboost
+
+#endif  // XGBOOST_TEST_MULTICLASS_OBJ_H_
```
Comment (icfaust) [inline tests/cpp/objective/test_multiclass_obj.h:19]: This is a nitpick but the end of file probably needs a change

Comment (razdoburdin) [reply tests/cpp/objective/test_multiclass_obj.h]: fixed

File: plugin/sycl/objective/multiclass_obj.cc [hunk 2]
```diff
@@ -1,30 +1,37 @@
...
+#pragma GCC diagnostic ignored "-Wtautological-constant-compare"
 #include "xgboost/data.h"
+#pragma GCC diagnostic pop
 #include "xgboost/logging.h"
 #include "xgboost/objective.h"
 #include "xgboost/json.h"
 
-#include "device_manager_oneapi.h"
+#include "../device_manager.h"
 #include "CL/sycl.hpp"
```
Comment (icfaust) [inline plugin/sycl/objective/multiclass_obj.cc:26]: Are we doing sycl2020 for this codebase yet?

File: plugin/sycl/objective/multiclass_obj.cc [hunk 3]
```diff
@@ -119,65 +126,53 @@ class SoftmaxMultiClassObjOneAPI : public ObjFunction {
...
+
+
+          bst_float const * point = &preds_acc[idx * nclass];
+
+
+          // Part of Softmax function
+          bst_float wmax = std::numeric_limits<bst_float>::min();
+          for (int k = 0; k < nclass; k++) { wmax = ::sycl::max(point[k], wmax); }
+          float wsum = 0.0f;
+          for (int k = 0; k < nclass; k++) { wsum += ::sycl::exp(point[k] - wmax); }
```
Comment (icfaust) [inline plugin/sycl/objective/multiclass_obj.cc:155]: for (int k = 0; k < nclass; k++) wsum += ::sycl::exp(point[k]);
sum *= ::sycl::exp(-wmax);

To reduce computation? comes at cost of possible overflow

File: plugin/sycl/objective/multiclass_obj.cc [hunk 4]
```diff
@@ -119,65 +126,53 @@ class SoftmaxMultiClassObjOneAPI : public ObjFunction {
...
+          if (label < 0 || label >= nclass) {
+            additional_input_acc[0] = 0;
+            label = 0;
+          }
+          bst_float wt = is_null_weight ? 1.0f : weights_acc[idx];
+          for (int k = 0; k < nclass; ++k) {
+            bst_float p = expf(point[k] - wmax) / static_cast<float>(wsum);
+            const float eps = 1e-16f;
+            const bst_float h = ::sycl::max(2.0f * p * (1.0f - p) * wt, eps);
+            p = label == k ? p - 1.0f : p;
```
Comment (icfaust) [inline plugin/sycl/objective/multiclass_obj.cc:166]: dumb question, p - 1 or 1-p here? It looks like because exp(point[k] - wmax) should be less than 1 or equal to 1.

File: plugin/sycl/objective/regression_obj.cc [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    qu_.submit([&](::sycl::handler& cgh) {
+        auto preds_acc            = preds_buf.get_access<::sycl::access::mode::read>(cgh);
+        auto labels_acc           = labels_buf.get_access<::sycl::access::mode::read>(cgh);
+        auto weights_acc          = weights_buf.get_access<::sycl::access::mode::read>(cgh);
+        auto out_gpair_acc        = out_gpair_buf.get_access<::sycl::access::mode::write>(cgh);
+        auto additional_input_acc = additional_input_buf.get_access<::sycl::access::mode::write>(cgh);
+        cgh.parallel_for<>(::sycl::range<1>(ndata), [=](::sycl::id<1> pid) {
+          int idx = pid[0];
+          bst_float p = Loss::PredTransform(preds_acc[idx]);
+          bst_float w = is_null_weight ? 1.0f : weights_acc[idx/n_targets];
```
Comment (icfaust) [inline plugin/sycl/objective/regression_obj.cc:99]: integer division here in the indexing. Usually that would want me to make sure I'm doing the right conversion. Worth a double check. Is ndata's idx per the 1st dimension. is idx supposed to be valued info.labels.Shape(1) spacing between value or is it like [1,2,3,4] etc.

File: plugin/sycl/objective/multiclass_obj.cc [hunk 6]
```diff
@@ -119,65 +126,53 @@ class SoftmaxMultiClassObjOneAPI : public ObjFunction {
...
+        auto additional_input_acc = additional_input_buf.template get_access<::sycl::access::mode::write>(cgh);
+        cgh.parallel_for<>(::sycl::range<1>(ndata), [=](::sycl::id<1> pid) {
+          int idx = pid[0];
+
+
+          bst_float const * point = &preds_acc[idx * nclass];
+
+
+          // Part of Softmax function
+          bst_float wmax = std::numeric_limits<bst_float>::min();
```
Comment (icfaust) [inline plugin/sycl/objective/multiclass_obj.cc:152]: Worth a quick check here: https://www.intel.com/content/www/us/en/developer/articles/technical/the-maxloc-reduction-in-oneapi.html#gs.07t59o to make sure we are following fastest practice here. Maybe a nitpick, maybe not.  If I were thinking how I'd want this to be offloaded to a gpu, I'm thinking how the logic would convert over, and how it may try to optimize.

**Review by icfaust — CHANGES_REQUESTED** [review]
So I don't have much of a clue on this repository. Some of my comments are probably very naive.

File: tests/cpp/plugin/test_sycl_regression_obj.cc [hunk 7]
```diff
@@ -4,131 +4,48 @@
...
+  TestSquaredLog(&ctx);
 }
 
 TEST(SyclObjective, LogisticRegressionGPair) {
   Context ctx;
   ctx.UpdateAllowUnknown(Args{{"device", "sycl"}});
-  std::vector<std::pair<std::string, std::string>> args;
-  std::unique_ptr<ObjFunction> obj { ObjFunction::Create("reg:logistic_oneapi", &ctx) };
-
-  obj->Configure(args);
```
[the remark this answers is not in the document]
Comment (razdoburdin) [reply tests/cpp/plugin/test_sycl_regression_obj.cc:66]: Currently we reuse the CPU tests. Just passed a `sycl `context to the `TestLogisticRegressionGPair`

File: plugin/sycl/tree/param.h [hunk 8]
```diff
@@ -14,25 +14,28 @@
 
 #include "xgboost/parameter.h"
 #include "xgboost/data.h"
+#pragma GCC diagnostic push
```
[the remark this answers is not in the document]
Comment (razdoburdin) [reply plugin/sycl/tree/param.h:17]: When compiled with `icpx `this #include generate numerous warning. Since this files are not a part of plugin, I suspended them.

File: plugin/sycl/objective/multiclass_obj.cc (hunk 3 above)
Comment (razdoburdin) [reply plugin/sycl/objective/multiclass_obj.cc:155]: Sounds reasonable. I guess in the future we could try to use this optimization.

**Review by ahuber21 — APPROVED** [review]
